### PR TITLE
Implement ci backend listener daemon

### PR DESCRIPTION
**Pull request by terceiro**
The listener daemon starts one child process for each backend, which will run
the listen() method in the corresponding backend implementation object. Changes
to the backend in database are detected, and listening processes are
started/restarted/stopped accordingly.

### CODE REVIEW

File: squad/ci/backend/lava.py [hunk 1]
```diff
@@ -60,19 +60,18 @@ def fetch(self, test_job):
             return self.__parse_results__(data)
 
     def listen(self):
-        if not self.data.listener_url:
-            self.log_debug("no listener_url, cannot listen for events")
-            return
+        url = urlsplit(self.data.url)
+        hostname = url.netloc
+
+        socket = self.__get_publisher_event_socket__()
```
Comment (mwasilew) [inline squad/ci/backend/lava.py]: This is returned in the full URL form, so I think we might just leave it as is.

File: squad/ci/models.py [hunk 2]
```diff
@@ -20,11 +20,6 @@ def list_backends():
 class Backend(models.Model):
     name = models.CharField(max_length=128, unique=True)
     url = models.URLField()
-    # listener_url is used by backend's listen() method
-    listener_url = models.URLField(null=True, blank=True)
```
Comment (mwasilew) [inline squad/ci/models.py]: I wouldn't go that far. LAVA offers a discovery method. Other services might not. Of course the problem is just 'theoretical' right now as we don't support other services for listening.

File: squad/ci/management/commands/listen.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            process = self.__processes__.get(backend.id)
+            if process:
+                # already running: restart if needed
+                if fields(backend) != fields(process.backend):
+                    self.stop(backend.id)
+                    self.start(backend)
+            else:
+                # not running, just start
+                self.start(backend)
+            if backend.id in ids:
```
Comment (mwasilew) [inline squad/ci/management/commands/listen.py]: shouldn't this be moved up to line 57? Looks like race condition. `start(backend)` will add ID to the `__processes__` dict and this block will remove it.

File: squad/ci/backend/__init__.py [hunk 4]
```diff
@@ -5,18 +5,19 @@
 from importlib import import_module
 
 
-__ALL_BACKENDS__ = OrderedDict()
+__ALL_BACKENDS__ = []
```
Comment (mwasilew) [inline squad/ci/backend/__init__.py]: +1. This was causing issues when trying to do fetch.delay (in next commit).

File: squad/ci/management/commands/listen.py (hunk 3 above)
Comment (terceiro) [inline squad/ci/management/commands/listen.py]: there is no race condition: `ids` is initialized as a _copy_ of `self.__processes__.keys()` so they are different lists.

File: squad/ci/models.py (hunk 2 above)
Comment (terceiro) [inline squad/ci/models.py]: I would prefer re-adding it later than keeping it there now without any function

File: squad/ci/backend/lava.py (hunk 1 above)
Comment (terceiro) [inline squad/ci/backend/lava.py]: yes but the hostname comes as `*`. what I can do is parse the url, replace the hostname if it's '*', and re-join the url.

Comment (terceiro) [inline squad/ci/backend/lava.py]: see https://bugs.linaro.org/show_bug.cgi?id=2993

what do you think?

File: squad/ci/models.py (hunk 2 above)
Comment (mwasilew) [inline squad/ci/models.py]: ACK

File: squad/ci/backend/lava.py (hunk 1 above)
Comment (mwasilew) [inline squad/ci/backend/lava.py]: I only checked with LKFT and it returned the proper address :( Anyway, I think your solution is OK. Just need to take the protocol/scheme from the reply to form the final address.

File: squad/ci/management/commands/listen.py (hunk 3 above)
Comment (mwasilew) [inline squad/ci/management/commands/listen.py]: OK

Comment (mwasilew) [inline squad/ci/management/commands/listen.py]: OK

**Comment by terceiro** [comment]
ping

**Pull request merged**